src/compiler/transformers/es2017.ts

```diff
@@ -83,6 +85,105 @@ namespace ts {
             }
         }
 
+        function asyncBodyVisitor(node: Node): VisitResult<Node> {
+            switch (node.kind) {
+                case SyntaxKind.VariableStatement:
+                    return visitVariableStatementInAsyncBody(<VariableStatement>node);
+                case SyntaxKind.ForStatement:
+                    return visitForStatementInAsyncBody(<ForStatement>node);
+                case SyntaxKind.ForInStatement:
+                    return visitForInStatementInAsyncBody(<ForInStatement>node);
+                case SyntaxKind.ForOfStatement:
+                    return visitForOfStatementInAsyncBody(<ForOfStatement>node);
+                case SyntaxKind.CatchClause:
+                    return visitCatchClauseInAsyncBody(<CatchClause>node);
+                case SyntaxKind.Block:
+                case SyntaxKind.SwitchStatement:
+                case SyntaxKind.CaseBlock:
+                case SyntaxKind.CaseClause:
+                case SyntaxKind.DefaultClause:
+                case SyntaxKind.TryStatement:
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.WhileStatement:
+                case SyntaxKind.IfStatement:
+                case SyntaxKind.WithStatement:
+                case SyntaxKind.LabeledStatement:
+                    return visitEachChild(node, asyncBodyVisitor, context);
+            }
+            Debug.assert(!isNodeWithPossibleVarDeclaration(node), "Unhandled node.");

```

Comment:
Why not make `isNodeWithPossibleVarDeclaration` into a type guard, start the function with it, the end with an `assertNever` to get the type checker to help out here?: 
```ts
if(!isNodeWithPossibleVarDeclaration(node)) {
  return visitor(node);
}
// ... switch block
Debug.assertNever(node, "Unhandled node.");
```